pkg/nats/stream/producer_client.go

```diff
@@ -0,0 +1,204 @@
+		CreatedAt: time.Now(),
+	}
+
+	if pc.activeStreamInfo[connDetails.ConnID] == nil {
+		pc.activeStreamInfo[connDetails.ConnID] = make(map[string]StreamInfo)
+	}
+
+	if _, ok := pc.activeStreamInfo[connDetails.ConnID][connDetails.StreamID]; ok {
+		return fmt.Errorf("cannot create request with same streamId %s again", connDetails.StreamID)
+	}
+
+	pc.activeStreamInfo[connDetails.ConnID][connDetails.StreamID] = streamInfo
+	pc.activeConnHeartBeatRequestSubjects[connDetails.ConnID] = connDetails.HeartBeatRequestSub
+
+	return nil
+}
+
+func (pc *ProducerClient) RemoveConnDetails(connDetails *ConnectionDetails) {
+	pc.mu.Lock()
+	defer pc.mu.Unlock()
+
+	activeStreamIdsForConn, ok := pc.activeStreamInfo[connDetails.ConnID]
+	if !ok {
+		return
+	}
+
+	if _, ok := activeStreamIdsForConn[connDetails.StreamID]; !ok {
+		return
+	}
+
+	delete(activeStreamIdsForConn, connDetails.StreamID)
+
+	if len(activeStreamIdsForConn) == 0 {
+		delete(pc.activeStreamInfo, connDetails.ConnID)
+		delete(pc.activeConnHeartBeatRequestSubjects, connDetails.ConnID)
+	}
+
+	if len(pc.activeStreamInfo) == 0 && pc.heartBeatCancelFunc != nil {
+		pc.heartBeatCancelFunc()
+		pc.heartBeatCancelFunc = nil
+	}
+}
+
+func (pc *ProducerClient) heartBeat(ctx context.Context) {
+	ctxWithCancel, cancel := context.WithCancel(ctx)
+	pc.heartBeatCancelFunc = cancel
+
+	ticker := time.NewTicker(pc.config.HeartBeatConfig.HeartBeatIntervalDuration)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-ctxWithCancel.Done():
+			log.Ctx(ctxWithCancel).Debug().Msgf("Heart beat for producer client %s", pc.ID)
+			return
+		case <-ticker.C:
+
+			nonActiveStreamIds := make(map[string][]string)
+
+			for c, v := range pc.activeConnHeartBeatRequestSubjects {
+				// Create an empty slice for activeStreamIds
+				var activeStreamIds []string
+
+				if streamInfoMap, ok := pc.activeStreamInfo[c]; ok {
+					for _, streamInfo := range streamInfoMap {
+						activeStreamIds = append(activeStreamIds, streamInfo.ID)
+					}
+				}
+
+				heartBeatRequest := HeartBeatRequest{
+					ProducerConnID:  pc.ID,
+					ActiveStreamIds: activeStreamIds,
+				}
+
+				data, err := json.Marshal(heartBeatRequest)
+				if err != nil {
+					log.Err(err)
+					continue
+				}
+
+				msg, err := pc.Conn.Request(v, data, pc.config.HeartBeatConfig.HeartBeatRequestTimeout)
+				if err != nil {
+					nonActiveStreamIds[c] = []string{}
+					continue
+				}
+
+				var heartBeatResponse ConsumerHeartBeatResponse
+				err = json.Unmarshal(msg.Data, &heartBeatResponse)
+				if err != nil {
+					continue
+				}
+				nonActiveStreamIds[c] = heartBeatResponse.NonActiveStreamIds
+			}
+
+			pc.updateActiveStreamInfo(nonActiveStreamIds)
+		}
+	}
+}
+
+func (pc *ProducerClient) WriteResponse(conn *ConnectionDetails, obj interface{}, writer *Writer) (int, error) {
+	pc.mu.Lock()
+	streamIds, active := pc.activeStreamInfo[conn.ConnID]
+
+	for _, v := range streamIds {
+		if v.ID == conn.StreamID {
+			active = true
+		}
+	}
```

Comment:
please revisit the choice of data structure to avoid looping over all streamIds for each write